torch_geometric/nn/models/revgnn.py

```diff
@@ -0,0 +1,288 @@
+import copy
+from abc import ABC, abstractmethod
+from typing import Optional, Union
+
+import numpy as np
+import torch
+from torch import Tensor, nn
+
+from torch_geometric.typing import Adj
+
+
+class InvertibleFunction(torch.autograd.Function):
+    """Invertible autograd function."""
+    @staticmethod
+    def forward(ctx, fn, fn_inverse, num_bwd_passes, num_inputs,
+                *inputs_and_weights):
+        ctx.fn = fn
+        ctx.fn_inverse = fn_inverse
+        ctx.weights = inputs_and_weights[num_inputs:]
+        ctx.num_bwd_passes = num_bwd_passes
+        ctx.num_inputs = num_inputs
+        inputs = inputs_and_weights[:num_inputs]
+        ctx.input_requires_grad = []
+
+        with torch.no_grad():
+            # Make a detached copy which shares the storage
+            x = []
+            for element in inputs:
+                if isinstance(element, torch.Tensor):
+                    x.append(element.detach())
+                    ctx.input_requires_grad.append(element.requires_grad)
+                else:
+                    x.append(element)
+                    ctx.input_requires_grad.append(None)
+            outputs = ctx.fn(*x)
+
+        if not isinstance(outputs, tuple):
+            outputs = (outputs, )
+
+        # Detaches outputs in-place, allows discarding the intermedate result
+        detached_outputs = tuple(element.detach_() for element in outputs)
+
+        # Clear memory of node features
+        inputs[0].storage().resize_(0)
+
+        # Store these tensor nodes for backward passes
+        ctx.inputs = [inputs] * num_bwd_passes
+        ctx.outputs = [detached_outputs] * num_bwd_passes
+
+        return detached_outputs
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
```

Comment:
Probably a good idea to test `backward` pass as well?

Reply:
Added test